File: 0410-split-array-largest-sum/0410-split-array-largest-sum.cpp
class Solution {
public:
    int fun(vector<int>& nums, int mid){
        int last = 0 , pos = 1 ;
        for(int i = 0 ; i < nums.size() ; i++){
            if(last + nums[i] <=mid)
                last += nums[i];
            else{
                last = nums[i];
                pos++;
            }
        }
        return pos;
    }
    int splitArray(vector<int>& nums, int k) {
        int low = *max_element(nums.begin(),nums.end());
        int high = accumulate(nums.begin(),nums.end(),0);

        while(low <= high){
            int mid = low +  (high-low)/2 ;
            int num = fun(nums,mid);
            if(num<=k)
                high = mid - 1 ;
            else
                low = mid + 1 ;
        }

        return low;    
    }
};

File: 0410-split-array-largest-sum/0410-split-array-largest-sum.java
class Solution {
    public boolean isPossible(int[] nums, int k, int mid){
        int sum = 0 ;
        int check = 1;
        for(int num : nums){
            if(num > mid)   return false ;
            if(sum + num > mid){
                check++ ;
                sum = num ;
                if(check > k)   return false;
            }
            else
                sum += num ;
        }
        return true ;
    }
    public int splitArray(int[] nums, int k) {
        int low = 0 ;
        int high = 0 ;
        int ans = -1 ;
        for(int num : nums){
            high += num ;
        }   

        while(low <= high){
            int mid = low + (high - low)/2 ;
            if(isPossible(nums,k,mid)){
                ans = mid ;
                high = mid - 1;
            }
            else
                low = mid + 1 ;

        }
        return ans ;
    }
}